yxi/axi-calyx/axi-combined-calyx.futil

```diff
@@ -0,0 +1,1239 @@
+      }
+
+
+      //txn bookkeeping.
+      //We are done performing reads when txn_count == txn_n
+      group txn_count_init {
+          txn_count.in = 32'b0;
+          txn_count.write_en = 1'b1;
+          txn_count_init[done] = txn_count.done;
+          
+      }
+
+      group txn_len_init {
+          // TODO(nathanielnrn):Parametrize this.
+          // 7 is good for word wide data bus and 8 element vec_add We'd
+          // expect 8 transfers. Number of transfers that occur is ARLEN + 1
+          txn_len.in = 8'd7;
+          txn_len.write_en = 1'b1;
+          txn_len_init[done] = txn_len.done;
+      }
+
+      group txn_incr {
+          txn_adder.left = txn_count.out;
+          txn_adder.right = 32'b1;
+          txn_count.in = txn_adder.out;
+          txn_count.write_en = 1'b1;
+          txn_incr[done] = txn_count.done;
+
+      }
+
+      //txn_count == txn_n
+      comb group check_reads_done {
+          perform_reads.left = txn_count.out;
+          perform_reads.right = txn_n.out;
+      }
+  }
+
+  control{
+      //XXX(nathanielnrn): What is best way to offer more flexiblity beyond just a counter?
+      seq{
+          txn_count_init;
+          txn_len_init;
+          while perform_reads.out with check_reads_done{
+              seq{
+                  par {
+                      reset_bt;
+                      reset_was_high;
+                  }
+                  do_ar_transfer;
+                  deassert_val;
+                  txn_incr;
+              }
+          }
+      }
+  }
+}
+
+
+
+
+component m_read_channel(
+  ARESET : 1,
+  RVALID : 1,
+  RLAST  : 1,
+  RDATA  : 32, 
+  RRESP :  2,  // Note: This is generated in subordinate! had this backwards in earlier version
+) -> (
+  // NOTE: In general, according to ZipCPU we want xREADY signals to be registered
+  // because (IIRC) it helps avoid combinational loops between READY and VALID.
+  RREADY : 1,
+) {
+  cells {
+      // 16 is due to dot-product vector length assumption
+      // For this manual implementation we are just writing into this data based
+      // on the data we read from cocotb
+      ref data_received = seq_mem_d1(32, 8, 64);
+      is_rdy = std_reg(1);
+      ref curr_addr = std_reg(64); 
+      
+      // registered because RLAST is high with last transfer, not after
+      // before this was registered we were terminating immediately with
+      // last transfer and not servicing it
+      n_RLAST = std_reg(1);
+
+      // TODO: get this width from yxi
+      read_data_reg = std_reg(32);
+
+      //address of seq_d1_mem we are writing to
+      curr_addr_adder = std_add(64);
+
+      // block_transfer reg to avoid combinational loops
+      bt_reg = std_reg(1);
+
+  }
+  wires{
+
+      RREADY = is_rdy.out;
+      data_received.read_en = 1'b0;
+
+      group init_n_RLAST {
+          n_RLAST.in = 1'b1;
+          n_RLAST.write_en = 1'b1;
+          init_n_RLAST[done] = n_RLAST.done;
+      }
```

Comment:
It's possible that simple register writes like this could be expressed with `invoke`s:

```
invoke n_RLAST(in=1'b1);
```

Reply:
I might be missing something about how `invoke` works. My understanding was that an `invoke` runs the control sequence of an instance. Because `std_reg` is defined as [verilog](https://github.com/calyxir/calyx/blob/94dd5b031b2c00d7731518cfd1f380db4fbd3382/primitives/compile.futil#L21-L43) what would invoking on a `std_reg` do?
Does `@go write-en: 1` mean that the `go` signal is tied to `write_en`? If so it makes sense that invoking a reg would drive `write_en` and we'd only need to supply a data input.

I could also imagine a world where we would also need to drive `write_en` explicitly:
```
invoke n_RLASt(in=1'b1, write_en=1'b1);
```

Reply:
Either way, I will address this here & in the generator in a future "simplifications" PR

Reply:
Yeah, to slightly generalize that understanding of `invoke`: it delegates control to the invoked cell and lets it run to completion. That actually works for both types of components (primitives and non-primitives), but it means different things mechanically: for primitives, it means "follow the go/done protocol to do an invocation of the Verilog module"; for non-primitives, it means "run the Calyx component's control program".

> Does `@go write-en: 1` mean that the `go` signal is tied to `write_en`?

Yes, exactly! The `@go` and `@done` annotations can be attached to any signal; they then play that role in the go/done interface.

So you don't need to supply `write_en` in the `invoke` because that's implicit in the invocation "calling convention."

Reply:
That's very useful! I'll test these changes out and probably push them to #1846